### PR TITLE
Added example regarding weight_decay distinction with per-parameter API

### LINKED ISSUE #115935: Penalisation of bias and BatchNorm parameters
### 📚 The doc issue

Currently, the most common example given when introducing optimisers[[1]](https://pytorch.org/tutorials/beginner/basics/optimization_tutorial.html) is something like:
```Python
optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)
```
This is of course a good first example, but perhaps we should consider that many users don't realise that `.parameters()` also contains the biases of the layers and the BatchNorm parameters, as it can be seen in #2639 , #1402 ,  [[2]](https://discuss.pytorch.org/t/weight-decay-in-the-optimizers-is-a-bad-idea-especially-with-batchnorm/16994/16), [[3]](https://discuss.pytorch.org/t/weight-decay-only-for-weights-of-nn-linear-and-nn-conv/114348). 

From what I understand of the current state of research [[4]](https://discuss.pytorch.org/t/weight-decay-in-the-optimizers-is-a-bad-idea-especially-with-batchnorm/16994/16), it's unclear whether penalising these parameters is a good practice or not, since these parameters should not induce overfitting.

I think it's a problem that this issue is not addressed anywhere in the documentation, as it should be clear that the default behaviour and examples adopt a practice that is not validated by current research.

### Suggest a potential alternative/fix

I believe that this issue should be addressed somewhere, possibly in the [torch.optim](https://pytorch.org/docs/stable/optim.html#module-torch.optim) page, maybe by specifying it [here](https://pytorch.org/docs/stable/optim.html#constructing-it).

In the same page we could change the provided example for Per-Parameter options to one in which the weight decay is used, as this is the current recommended way in which one should address the separation of parameter penalties. _Could be_ something like:
```python
penalized     = [p for name,p in self.named_parameters() if 'bias' not in name]
not_penalized = [p for name,p in self.named_parameters() if 'bias' in name]
optimizer_params = [
    {'params': penalized, 'weight_decay':1e-3,  'lr'=1e-3},
    {'params': not_penalized, 'weight_decay':0}
]
optim.SGD(optimizer_params, lr=1e-2, momentum=0.9)
```

Also I believe it may be a nice idea to include this in the [optimization tutorial](https://pytorch.org/tutorials/beginner/basics/optimization_tutorial.html) somewhere near the end.

-------
This is my first contribution here on pytorch, I hope that this issue is understable, respects the contributing guidelines and it's relevant. In case this change would be welcome I would be glad to implement this myself.

cc @svekars @brycebortree @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

### PR DESCRIPTION
**Pull request by Mamiglia**
Added new example and description regarding per-parameter `weight_decay` distinction for bias and non-bias terms. 

Fixes #115935

### CODE REVIEW

**Comment by Mamiglia** [comment]
I also moved a note down a paragraph as it seemed better positioned in that way.

File: docs/source/optim.rst [hunk 1]
```diff
@@ -39,16 +50,23 @@ as optimization options for this group.
     between parameter groups.
 
 
-For example, this is very useful when one wants to specify per-layer learning rates::
+Consider the following example related to the distinct penalization of parameters.
```
Comment (janeyx99) [inline docs/source/optim.rst:53]:
```suggestion
Also consider the following example related to the distinct penalization of parameters.
```
nit, just adding a transitional word

File: docs/source/optim.rst [hunk 2]
```diff
@@ -39,16 +50,23 @@ as optimization options for this group.
...
+Remember that :func:`~torch.nn.Module.parameters` returns an iterable that 
+contains all learnable parameters, including biases and other 
+parameters that may need distinct penalization. To address this, one can specify
```
Comment (janeyx99) [inline docs/source/optim.rst:56]:
```suggestion
parameters that may prefer distinct penalization. To address this, one can specify
```

File: docs/source/optim.rst [hunk 3]
```diff
@@ -39,16 +50,23 @@ as optimization options for this group.
...
+individual penalization weights for each parameter group::
+
+    bias_params = [p for name,p in self.named_parameters() if 'bias' in name]
+    others = [p for name,p in self.named_parameters() if 'bias' not in name]
```
Comment (janeyx99) [inline docs/source/optim.rst:60]:
```suggestion
    bias_params = [p for name, p in self.named_parameters() if 'bias' in name]
    others = [p for name, p in self.named_parameters() if 'bias' not in name]
```

File: docs/source/optim.rst [hunk 4]
```diff
@@ -39,16 +50,23 @@ as optimization options for this group.
...
 
     optim.SGD([
-                    {'params': model.base.parameters()},
-                    {'params': model.classifier.parameters(), 'lr': 1e-3}
-                ], lr=1e-2, momentum=0.9)
+                    {'params': others },
+                    {'params': bias_params, 'weight_decay': 0}
+                ], weight_decay=1e-3, lr=1e-2)
```
Comment (janeyx99) [inline docs/source/optim.rst:65]: 1e-3 is a really small weight_decay--is this intentional?

File: docs/source/optim.rst [hunk 5]
```diff
@@ -31,6 +31,17 @@ a ``params`` key, containing a list of parameters belonging to it. Other keys
...
 
+For example, this is very useful when one wants to specify per-layer learning rates::
+
+    optim.SGD([
+                    {'params': model.base.parameters()},
+                    {'params': model.classifier.parameters(), 'lr': 1e-3}
+                ], lr=1e-2, momentum=0.9)
+
+This means that ``model.base``'s parameters will use the default learning rate of ``1e-2``,
+``model.classifier``'s parameters will use a learning rate of ``1e-3``, and a momentum of
```
Comment (janeyx99) [inline docs/source/optim.rst:42]:
```suggestion
    optim.SGD([
                    {'params': model.base.parameters(), 'lr': 1e-2},
                    {'params': model.classifier.parameters()}
                ], lr=1e-3, momentum=0.9)

This means that ``model.base``'s parameters will use a learning rate of ``1e-3``,
``model.classifier``'s parameters will use the default learning rate of ``1e-3``, and a momentum of
```

Might need some spacing updates, but let's also change the phrasing here to be more consistent with expectation. Recently, we determined SGD's default LR to be 1e-3, so making that the "default" in this case will raise fewer eyebrows.

File: docs/source/optim.rst [hunk 6]
```diff
@@ -39,16 +50,23 @@ as optimization options for this group.
...
-                    {'params': model.classifier.parameters(), 'lr': 1e-3}
-                ], lr=1e-2, momentum=0.9)
+                    {'params': others },
+                    {'params': bias_params, 'weight_decay': 0}
+                ], weight_decay=1e-3, lr=1e-2)
 
-This means that ``model.base``'s parameters will use the default learning rate of ``1e-2``,
-``model.classifier``'s parameters will use a learning rate of ``1e-3``, and a momentum of
-``0.9`` will be used for all parameters.
+In this manner, bias terms are isolated from non-bias terms, and a ``weigth_decay``
```
Comment (janeyx99) [inline docs/source/optim.rst:67]:
```suggestion
In this manner, bias terms are isolated from non-bias terms, and a ``weight_decay``
```

**Review by janeyx99 — COMMENTED** [review]
Thanks for the quick change!

Let some nits--could you also paste a screenshot of the docs preview of these changes to ensure expected formatting? (The docs preview can be accessed through the link to the Python docs in the big pytorch-bot comment.)

File: docs/source/optim.rst (hunk 4 above)
Comment (Mamiglia) [reply docs/source/optim.rst:65]: No it's completely random. Can be changed if you believe it's too small

Comment (Mamiglia) [reply docs/source/optim.rst:65]: Actually, just to be sure i'll change it to `1e-2`.

**Comment by Mamiglia** [comment]
I'm not sure if it's because there's an issue with the formatting/syntax but the link gives me this everytime. I'll look into this.

```
<Error>
<Code>NoSuchKey</Code>
<Message>The specified key does not exist.</Message>
<Key>pytorch/pytorch/117436/index.html</Key>
<RequestId>TKCF10WQRQX7G68V</RequestId>
<HostId>
dCRA4QyKsnThfUI2RYPtYxJG8J3+BjbkWynJpGiWXWYYh9d56dKGuq4SfVBq4sBj04B58Aq2Ljs=
</HostId>
</Error>
```

**Comment by Mamiglia** [comment]
It appears it was an unneeded newline at the end of the file. This is the final result.

image

image

**Comment by janeyx99** [comment]
> I'm not sure if it's because there's an issue with the formatting/syntax but the link gives me this everytime. I'll look into this.
> 
> ```
> <Error>
> <Code>NoSuchKey</Code>
> <Message>The specified key does not exist.</Message>
> <Key>pytorch/pytorch/117436/index.html</Key>
> <RequestId>TKCF10WQRQX7G68V</RequestId>
> <HostId>
> dCRA4QyKsnThfUI2RYPtYxJG8J3+BjbkWynJpGiWXWYYh9d56dKGuq4SfVBq4sBj04B58Aq2Ljs=
> </HostId>
> </Error>
> ```

ah yea this happens when the workflow that builds the docs isn't done yet so the preview hasn't been uploaded.

Thanks for your changes! This looks good, will confirm once the docs build is done before approving.

File: docs/source/optim.rst [hunk 7]
```diff
@@ -31,24 +31,42 @@ a ``params`` key, containing a list of parameters belonging to it. Other keys
...
 
-For example, this is very useful when one wants to specify per-layer learning rates::
+    bias_params = [p for name, p in self.named_parameters() if 'bias' in name]
+    others = [p for name, p in self.named_parameters() if 'bias' not in name]
 
     optim.SGD([
-                    {'params': model.base.parameters()},
-                    {'params': model.classifier.parameters(), 'lr': 1e-3}
-                ], lr=1e-2, momentum=0.9)
+                    {'params': others },
```
Comment (janeyx99) [inline docs/source/optim.rst:62]:
```suggestion
                    {'params': others},
```
nit

**Review by janeyx99 — APPROVED** [review]

File: docs/source/optim.rst (hunk 7 above)
Comment (janeyx99) [reply docs/source/optim.rst:62]: Once you commit this, feel free to ping me so i can help land the PR.

**Comment by Mamiglia** [comment]
Done! I hope everything fits right! @janeyx99

File: docs/source/optim.rst (hunk 4 above)
Comment (janeyx99) [reply docs/source/optim.rst]: people normally use 0.1 i believe btw

**Comment by Mamiglia** [comment]
I see that there're some tests failing. I tried to inspect what's the issue there, but I couldn't understand it. It seems unrelated with the changes in the doc

**Comment by janeyx99** [comment]
Yes, the bot has identified these as flaky
image
